src/SonarScanner.MSBuild.PreProcessor/Roslyn/RoslynRuleSetGenerator.cs

```diff
@@ -56,10 +57,12 @@ public RuleSet Generate(IEnumerable<SonarRule> activeRules, IEnumerable<SonarRul
                 throw new ArgumentNullException(nameof(language));
             }
 
-            var allRulesByPartialRepoKey = RoslynRulesByPartialRepoKey(activeRules, language)
-                .Concat(RoslynRulesByPartialRepoKey(inactiveRules, language))
-                .GroupBy(x => x.Key)
-                .ToDictionary(x => x.Key, x => x.SelectMany(y => y.Value));
+            var rulesElements = activeRules.Concat(inactiveRules)
+                .GroupBy(
+                    rule => GetPartialRepoKey(rule, language),
+                    rule => rule)
+                .Where(IsSupportedRuleRepo)
+                .Select(CreateRulesElement);
```

Comment:
Replacing complex logic with GroupBy. This should be functionally equivalent, but pay attention here when reviewing.

Reply:
Looks ok. How well is this covered by unit tests?

Reply:
I added several tests in my last commits, should be covered better now.